calico-vpp-agent/policy/policy_server.go

```diff
@@ -1098,21 +1103,181 @@ func (s *Server) handleHostMetadataRemove(msg *proto.HostMetadataRemove, pending
 }
 
 func (s *Server) handleIpamPoolUpdate(msg *proto.IPAMPoolUpdate, pending bool) (err error) {
-	common.SendEvent(common.CalicoVppEvent{
-		Type: common.IpamPoolUpdate,
-		New:  msg,
-	})
+	if msg != nil {
+		err = s.handleIPPoolUpdate(msg, nil)
+		if err != nil {
+			return errors.Wrap(err, "error processing pool add / modified")
+		}
+	}
 	return nil
 }
 
 func (s *Server) handleIpamPoolRemove(msg *proto.IPAMPoolRemove, pending bool) (err error) {
+	if msg != nil {
+		err = s.handleIPPoolUpdate(nil, msg)
+		if err != nil {
+			return errors.Wrap(err, "error processing pool del")
+		}
+	}
+	return nil
+}
+
+// update updates the internal map with IPAM updates when the update
+// is new addtion to the map or changes the existing item, it calls
+// ipamUpdateHandler
+func (s *Server) handleIPPoolUpdate(poolUpdate *proto.IPAMPoolUpdate, poolRemove *proto.IPAMPoolRemove) error {
+	s.ippoolLock.Lock()
+	defer s.ippoolLock.Unlock()
+	key := ""
+	if poolUpdate != nil {
+		key = poolUpdate.Id
+	} else {
+		key = poolRemove.Id
+	}
+
+	if key == "" {
+		s.log.Debugf("Empty pool")
+		return nil
+	}
+
+	existing, found := s.ippoolmap[key]
+	if poolRemove != nil {
```

Comment:
We can probably make keep the two functions separated, the shared logic is mostly the `ippoolmap`